apps/api/src/app/workflows-v2/usecases/get-step-schema/get-step-metadata-use-case.service.ts

```diff
@@ -0,0 +1,150 @@
+import { BadRequestException, Injectable } from '@nestjs/common';
+import { ControlValuesLevelEnum, StepMetadataDto, StepType, StepTypeEnum } from '@novu/shared';
+import { JSONSchema } from 'json-schema-to-ts';
+import { ControlValuesRepository, NotificationStepEntity, NotificationTemplateRepository } from '@novu/dal';
+import { GetStepSchemaCommand } from './get-step-schema.command';
+import { mapStepTypeToResult } from '../../shared';
+
+@Injectable()
+export class GetStepMetadataUseCase {
+  constructor(
+    private notificationTemplateRepository: NotificationTemplateRepository,
+    private controlValuesRepository: ControlValuesRepository
+  ) {}
+
+  async execute(command: GetStepSchemaCommand): Promise<StepMetadataDto> {
+    const { currentStep, previousSteps } = await this.findSteps(command);
+    this.assertStepValidity(currentStep);
+
+    return {
+      controls: {
+        dataSchema: currentStep.controls?.schema,
+        uiSchema: currentStep.controls?.uiSchema,
+        values: await this.getValues(command, currentStep),
+      },
+      variables: buildVariablesSchema(previousSteps),
+    };
+  }
+
+  private async getValues(command: GetStepSchemaCommand, currentStep: NotificationStepEntity) {
+    const controlValuesEntity = await this.controlValuesRepository.findFirst({
```

Comment:
We have to be stricter with the database and assume that there will only be a single entry in the controls collection.  That is, instead of findFirst we should use findOne.